In-progress reworking of product subscription page for new payment flows

lmorchard opened the pull request:
Starting a draft of this PR for issue #5745. Lots of work left to do, but worth starting to take a look at.

Following along with [Stripe's fixed-price subscription example](https://stripe.com/docs/billing/subscriptions/fixed-price), we need to handle a small flowchart of process here.

It's growing into a significant re-write of the product subscription page. And, some of what's here might end up useful to reuse for using the endpoints from #5744 in updating payment method.

lmorchard commented on packages/fxa-payments-server/src/App.tsx:
```diff
@@ -22,6 +22,7 @@ import * as FlowEvents from './lib/flow-event';
 import { observeNavigationTiming } from './lib/navigation-timing';
 
 const Product = React.lazy(() => import('./routes/Product'));
+const ProductV2 = React.lazy(() => import('./routes/ProductV2'));
```
Started by copying `routes/Product` into `routes/ProductV2` to start making changes to a parallel version.

lmorchard commented on packages/fxa-payments-server/src/lib/apiClient.ts:
```diff
@@ -257,3 +257,52 @@ export async function apiUpdatePayment(params: {
   }
 }
+
+export async function apiCreateCustomer(params: {
```
Adding new API client methods for new auth-server routes, while trying to roll in better type annotations.

lmorchard commented on packages/fxa-payments-server/src/routes/ProductV2/SubscriptionCreate/index.stories.tsx:
```diff
@@ -0,0 +1,143 @@
+import React from 'react';
```
These are all-new stories that I'm starting to build up to exercise the modes of the new payment flow. Still much in progress.

lmorchard commented on packages/fxa-payments-server/src/routes/ProductV2/SubscriptionCreate/index.tsx:
```diff
@@ -0,0 +1,472 @@
+import React, {
```
This is based on the `routes/Product/SubscriptionCreate` component, but getting a significant overhaul to accommodate the new retry and confirmation flows.

lmorchard commented on packages/fxa-payments-server/src/routes/ProductV2/SubscriptionCreate/index.tsx:
```diff
@@ -0,0 +1,472 @@
+  retryInvoiceInitialState,
+  apiClientOverrides = {},
+}: SubscriptionCreateProps) => {
+  const {
```
A big departure from the previous payment page: Dropping use of Redux in favor of our new `useAwait` hook and managing API usage here locally.

lmorchard commented on packages/fxa-payments-server/src/routes/ProductV2/SubscriptionCreate/index.tsx:
```diff
@@ -0,0 +1,472 @@
+
+  const [retryStatus, setRetryStatus] = useState<RetryStatus>(null);
+
+  const onSubmit: PaymentFormProps['onSubmit'] = useCallback(
```
The new PaymentFormV2 no longer handles `onPayment` and `onPaymentError` for token creation. The new `onSubmit` handler should hopefully pass along everything needed to handle payment processing here in SubscriptionCreate

lmorchard commented on packages/fxa-payments-server/src/routes/ProductV2/SubscriptionCreate/index.tsx:
```diff
@@ -0,0 +1,472 @@
+
+  const onSubmit: PaymentFormProps['onSubmit'] = useCallback(
+    (params) =>
+      handleSubscriptionPayment({
```
This is more verbose than I'd like, but I'm trying to pass along enough context to the `handleSubscriptionPayment` function so that it or its subroutines could possibly be extracted from this component and put in a shared module.

lmorchard commented on packages/fxa-payments-server/src/routes/ProductV2/SubscriptionCreate/index.tsx:
```diff
@@ -0,0 +1,472 @@
+  ) : null;
+
+export const handleSubscriptionPayment = async ({
+  stripe,
```
This is where the flow gets complicated. Following along with [Stripe's fixed-price subscription example](https://stripe.com/docs/billing/subscriptions/fixed-price), we need to handle a small flowchart of process here.

Not done yet, but trying to arrange this so that some of this could be extracted out of this component.

lmorchard commented on packages/fxa-payments-server/src/routes/ProductV2/SubscriptionSuccess/index.test.tsx:
```diff
@@ -0,0 +1,59 @@
+import React from 'react';
```
None of these tests will pass, yet.

lmorchard commented:
Decided to hack on this a bit more after dinner. Lots of tests left to write, but mostly have story coverage for various states.

Started tinkering on integration with the new auth-server endpoints. Things are a bit hacked up: 

* Ran into some issues with validators and reused idempotency keys. 
* Got things to the point where a subscription is successfully created. 
* But, it looks like the customer doesn't have any payment sources after the subscription is created this way. This results in an error on refreshing the customer object.

chenba commented on packages/fxa-auth-server/lib/payments/stripe.js:
```diff
@@ -300,7 +300,7 @@ class StripeHelper {
           customer: customerId,
         },
-        { idempotencyKey }
+        { idempotencyKey: `pma-${idempotencyKey}` }
```
What's the significance of these prefixes?

lmorchard replied:
The prefix is just an arbitrary string to make each of the keys unique. Reusing the same idempotency key for both of those API calls resulted in the second one failing. I forget the specific error message, but the gist was Stripe wants distinct idempotency keys for different API calls.

lmorchard commented:
Squashed and rebased against main. Current status:

* Seems to work end-to-end for creating a new subscription
* No errors on subscription management or download pages
* Need to write tests for the new UI code
* Payment details (i.e. last4, brand, exp date) are missing from customer object from the API, so that's still a wrinkle to iron out.

lmorchard commented:
The problems with fetching & displaying payment details are now split off into separate issues:

* API work to find payment details for this new subscription flow - https://github.com/mozilla/fxa/issues/5902
* UI work to accommodate missing payment details (e.g. for trials) - https://github.com/mozilla/fxa/issues/5901

That being the case, I'm going to wrap up tests for this PR and then hopefully open it up for proper review.

lmorchard commented:
Another rebase against main. Re-enabled validators with some tweaks and rough unit tests with copypasta API data that was failing validation.

bbangert commented on packages/fxa-auth-server/test/local/routes/validators.js:
```diff
@@ -450,4 +450,130 @@ describe('lib/routes/validators:', () => {
+      const res = subject.validate(data);
+      assert.ok(!res.error);
+    });
+  });
```
There should be fixtures for these.

lmorchard replied:
Should be - as in they already exist? Or should be newly created?

lmorchard commented:
Opening this for review, hopefully to wrap up early next week. Looks like I have 100% test coverage on the new route components, locally at least.

Rebased against main, squashed all the commits, now leaving it alone for the weekend.

chenba commented on packages/fxa-auth-server/lib/routes/validators.js:
```diff
@@ -470,17 +470,22 @@ module.exports.subscriptionsStripeSubscriptionItemValidator = isA
-    cancel_at: isA.number().optional(),
-    canceled_at: isA.number().optional(),
+    cancel_at: isA.alternatives(isA.number(), isA.any().allow(null)),
+    canceled_at: isA.alternatives(isA.number(), isA.any().allow(null)),
```
Isn't `isA.number()` a subset of `isA.any().allow(null)`?  I'm trying to understand why we'd need both.

lmorchard replied:
I don't think so - I think it's either a number or a null. The `any()` validator starts generic, but `allow()` restricts it to just null. We use this construction elsewhere in validators here

chenba commented on packages/fxa-auth-server/test/local/routes/validators.js:
```diff
@@ -450,4 +450,130 @@ describe('lib/routes/validators:', () => {
+
+  describe('subscriptionsStripeSubscriptionValidator', () => {
+    const { subscriptionsStripeSubscriptionValidator: subject } = validators;
+    it('accepts incomplete subscription with empty items and null cancel_at', () => {
```
Does "empty items" here in the description mean that `data.items` below should not have any objects in its list?  I do see an object.

lmorchard replied:
Yeah, this description is kind of a guess at what was failing in the validator. Turns out it wasn't an items data object with an empty list - the problem was that item is itself an object rather than a list.

chenba commented on packages/fxa-payments-server/src/components/fields/index.tsx:
```diff
@@ -252,7 +247,22 @@ export const defaultStripeElementValidator: OnValidateFunction = (
+} & FieldProps & {
+    component: any;
+    getString: Function | undefined;
+  };
```
`getString?: Function` instead?

Any reason to split up `{ onValidate? }` and `{ component; getString }`?

lmorchard commented:
Made a few tweaks from review comments. Rebased against main and restarted tests until they finally stopped failing on unrelated things 😥

chenba commented on packages/fxa-payments-server/src/routes/ProductV2/SubscriptionCreate/index.tsx:
```diff
@@ -0,0 +1,363 @@
+        paymentIntentClientSecret,
+        { payment_method: paymentMethodId }
+      );
+      console.log('CONFIRM RESULT', confirmResult);
```
Debug logging?

lmorchard replied:
Yeah, that's debug logging 😞

chenba commented on packages/fxa-payments-server/src/routes/ProductV2/SubscriptionCreate/index.tsx:
```diff
@@ -0,0 +1,363 @@
+    }
+    // Other payment_intent.status cases?
+    default: {
+      console.error('Unexpected payment intent status', paymentIntentStatus);
```
Debug logging?

lmorchard replied:
No, this is an actual error we'll want to see, if it happens. I think this gets picked up by sentry

chenba replied:
👍 Also just noticed the tests.

chenba commented on packages/fxa-payments-server/src/routes/ProductV2/SubscriptionCreate/index.test.tsx:
```diff
@@ -0,0 +1,633 @@
+  describe('errors', () => {
+    // We map Stripe error codes to less-specific error message string IDs for display.
+    // L10n is disabled for these error messages in testing, so they show up as string IDs.
+    // The codes are mapped to IDs in lib/errors.ts and the strings are in public/locales
```
These comments are very helpful.  Perhaps we should move/copy them to where errors mapped/handled.

lmorchard replied:
Well, my intent for the comment here is to explain why we're asserting for codes like `getByText('card-error')` rather than `getByText('Your transaction could not be processed. Please verify your credit card information and try again.
')` - because l10n is disabled here and that skips a couple levels of indirection

chenba commented on packages/fxa-payments-server/src/routes/ProductV2/SubscriptionCreate/index.test.tsx:
```diff
@@ -0,0 +1,633 @@
+      expect(screen.getByText('payment-error-2')).toBeInTheDocument();
+    });
+  });
+});
```
💯 ALL THE TESTS

lmorchard commented:
Going to try to get one more green CI run and then probably merge.

lmorchard commented:
One more rebase against main to see if I pick up some CI fixes

lmorchard commented:
Finally a passing CI run. Merging while I can!
